File: apps/web/src/types/hostel.ts
export enum HostelType {
    BOYS = "BOYS",
    GIRLS = "GIRLS",
    STAFF = "STAFF",
    GUEST = "GUEST"
}

export enum RoomType {
    AC = "AC",
    NON_AC = "NON_AC",
    SUITE = "SUITE"
}

export enum GatePassType {
    LOCAL = "LOCAL",
    HOME = "HOME",
    EMERGENCY = "EMERGENCY"
}

export enum GatePassStatus {
    PENDING = "PENDING",
    APPROVED = "APPROVED",
    REJECTED = "REJECTED",
    OUT = "OUT",
    RETURNED = "RETURNED",
    EXPIRED = "EXPIRED"
}

export enum ComplaintStatus {
    OPEN = "OPEN",
    IN_PROGRESS = "IN_PROGRESS",
    RESOLVED = "RESOLVED",
    CLOSED = "CLOSED"
}

export interface HostelBlock {
    id: number;
    name: string;
    type: HostelType;
    total_floors: number;
    warden_name?: string;
    contact_number?: string;
}

export interface HostelRoom {
    id: number;
    block_id: number;
    room_number: string;
    floor: number;
    capacity: number;
    current_occupancy: number;
    room_type: RoomType;
    monthly_rent: number;
    is_active: boolean;
}

export interface BedAllocation {
    id: number;
    room_id: number;
    bed_number: string;
    student_id?: number;
    faculty_id?: number;
    allocation_date: string;
    deallocation_date?: string;
    is_active: boolean;
}

export interface GatePass {
    id: number;
    student_id: number;
    type: GatePassType;
    out_time: string;
    expected_in_time: string;
    actual_in_time?: string;
    reason: string;
    status: GatePassStatus;
    approved_by?: number;
    remarks?: string;
}

export interface HostelComplaint {
    id: number;
    student_id: number;
    room_id: number;
    category: string;
    description: string;
    priority: string;
    status: ComplaintStatus;
    resolution_note?: string;
    resolved_at?: string;
    created_at: string;
}

File: apps/api/app/models/hostel.py
from enum import Enum
from typing import Optional, List, TYPE_CHECKING
from sqlmodel import SQLModel, Field, Relationship
from datetime import datetime

if TYPE_CHECKING:
    from .student import Student
    from .faculty import Faculty

class HostelType(str, Enum):
    BOYS = "BOYS"
    GIRLS = "GIRLS"
    STAFF = "STAFF"
    GUEST = "GUEST"

class RoomType(str, Enum):
    AC = "AC"
    NON_AC = "NON_AC"
    SUITE = "SUITE"

class GatePassType(str, Enum):
    LOCAL = "LOCAL"
    HOME = "HOME"
    EMERGENCY = "EMERGENCY"

class GatePassStatus(str, Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    OUT = "OUT"
    RETURNED = "RETURNED"
    EXPIRED = "EXPIRED"

class ComplaintStatus(str, Enum):
    OPEN = "OPEN"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    CLOSED = "CLOSED"

# --- Infrastructure ---

class HostelBlock(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(unique=True, index=True)
    type: HostelType
    total_floors: int
    warden_name: Optional[str] = None
    contact_number: Optional[str] = None
    
    rooms: List["HostelRoom"] = Relationship(back_populates="block")

class HostelRoom(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    block_id: int = Field(foreign_key="hostelblock.id")
    room_number: str = Field(index=True)
    floor: int
    capacity: int
    current_occupancy: int = Field(default=0)
    room_type: RoomType = Field(default=RoomType.NON_AC)
    monthly_rent: float
    is_active: bool = Field(default=True)

    block: HostelBlock = Relationship(back_populates="rooms")
    allocations: List["BedAllocation"] = Relationship(back_populates="room")

class BedAllocation(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    room_id: int = Field(foreign_key="hostelroom.id")
    bed_number: str
    student_id: Optional[int] = Field(default=None, foreign_key="student.id")
    faculty_id: Optional[int] = Field(default=None, foreign_key="faculty.id")
    allocation_date: datetime = Field(default_factory=datetime.utcnow)
    deallocation_date: Optional[datetime] = None
    is_active: bool = Field(default=True)
    
    room: HostelRoom = Relationship(back_populates="allocations")

# --- Operations ---

class GatePass(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    student_id: int = Field(foreign_key="student.id")
    type: GatePassType
    out_time: datetime
    expected_in_time: datetime
    actual_in_time: Optional[datetime] = None
    reason: str
    status: GatePassStatus = Field(default=GatePassStatus.PENDING)
    approved_by: Optional[int] = Field(default=None, foreign_key="user.id")
    remarks: Optional[str] = None

class HostelComplaint(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    student_id: int = Field(foreign_key="student.id")
    room_id: int = Field(foreign_key="hostelroom.id")
    category: str  # Plumbing, Electrical, Cleaning, etc.
    description: str
    priority: str = Field(default="MEDIUM") # LOW, MEDIUM, HIGH, URGENT
    status: ComplaintStatus = Field(default=ComplaintStatus.OPEN)
    resolution_note: Optional[str] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
